pkg/k8sclient/k8sclient.go

```diff
@@ -251,12 +251,17 @@ func parsePodNetworkAnnotation(podNetworks, defaultNamespace string) ([]*types.N
 	return networks, nil
 }
 
-func getKubernetesDelegate(client *ClientInfo, net *types.NetworkSelectionElement, confdir string, pod *v1.Pod, resourceMap map[string]*types.ResourceInfo) (*types.DelegateNetConf, map[string]*types.ResourceInfo, error) {
+func getKubernetesDelegate(client *ClientInfo, net *types.NetworkSelectionElement, confdir string, pod *v1.Pod, resourceMap map[string]*types.ResourceInfo, try bool) (*types.DelegateNetConf, map[string]*types.ResourceInfo, error) {
 
-	logging.Debugf("getKubernetesDelegate: %v, %v, %s, %v, %v", client, net, confdir, pod, resourceMap)
+	logging.Debugf("getKubernetesDelegate: %v, %v, %s, %v, %v, %v", client, net, confdir, pod, resourceMap, try)
 	customResource, err := client.NetClient.NetworkAttachmentDefinitions(net.Namespace).Get(context.TODO(), net.Name, metav1.GetOptions{})
 	if err != nil {
```

Comment:
This handles not only NotFound also serious error case, but in this case, we can change the behavior for NotFound error case and other error case as follows:

- suppress error to output kube events, print log with `logging.Verbose` for internal use and go further in NotFound case
- print error message to kube events (as current code does) in other error case

What do you think about that?

Reply:
Good catch, I missed the other errors situation, but I think for the other cases we should report the event when nad not found,won't that be a bigger change in behavior?

Reply:
As I noted, other cases (not NotFound case), we should report the error in event. Let me share my idea which is in above comment:

- If NotFound error, we don't report the error as Kubernetes event, however, we should report it in multus logging, with  `logging.Verbosef()` for further troubleshooting (we need to report this case we don't use net-attach-def)
- If other error, we should report in Kubernetes event as currently we did

Reply:
Almost OK to me. Just a double check that this new proposal will remove `try` argument from function, is that correct?

Reply:
Yes this would not have the 'try', but what about the behavior change in the other functions?

Reply:
Please let me know what is the behavior change without try. As far as I know of, this 'try' is tntroduced in your PR.

Reply:
We support CNI json file in pod network annotation, as well as defaultNetwork/clusterNetwork. Your PR seems to only support CNI file in defaultNetwork/clusterNetwork with `try`, so your PR changes the behavior.

We want to support both, so please remove `try` and keep current behavior.

Reply:
But in 511 and 673 we return right away when `getKubernetesDelegate` fails, in contrast to line 544 where we do not fail on err != nil and do check the CNI config as a failback. My `try` ensures we do not report error when this failback happens.

Reply:
I read your code again and I make sure that we need to report the error in both case, 511, 673. 

So please let me know the reason why you do not report the error?

Reply:
I read your code again and I make sure that we need to report the error in both case, 511, 673. 

The difference between 511 and 673 is retrieve CNI config for network annotation in pod, or 'clusterNetwork'/'defaultNetwork' config in multus config. Both case, we should report the error to users otherwise user may not recognize error. However, your PR intentionally skip to report the error in 'clusterNetwork'/'defaultNetworks' case, by adding `try` clause.
For example, if one of 'defaultNetworks' is not invalid (i.e. net-attach-def is not exist), your code does not show any error for that. Current multus code, of course, shows the error and let the user to know users' config has something wrong. Both case, we should report the error to Kubernetes events to easily understand what is happen. So I still believe we do not need your 'try' clause.

So please let me know the reason why you do not report the error in ?

Reply:
In addition, I am not clear why `try` does skipping its error report. Could you please explain why it names `try`?

Reply:
Because in 544 (where I set `try=true` and change the behavior) we failback to the CNI json check, and only if that fails report the error.

I probably could add event report at the end of that function (line 607) to ensure we eventually report the event on failure.

Reply:
> Because in 544 (where I set try=true and change the behavior) we failback to the CNI json check, and only if that fails report the error.

As Doug noted, net-attach-def is required even if CNI config name is in Pod annotation. In this case, net-attach-def does not have any config in spec, like following (that's why config is optional):

```
---
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: etc-cni-config1
spec:
```

Reply:
To put more light to my use case, my `00-multus.conf` is:
```
{
  "name": "multus-cni-network",
  "type": "multus",
  "confDir": "/etc/cni/net.d",
...
  "clusterNetwork": "cilium"
}
```

where I am using the `clusterNetwork`, but I do not create a NAD for it, and just rely on the existing cni conf for `cilium` inside the `/etc/cni/net.d` path.
In such setup, whenever I create default Pod, (w/o any additional networks) I will get an error event that cilium NAD not found, where in reality all works.

What if we keep the `try`, and I expand the `if` with `apierrors.IsNotFound(err)` and add logging, like this:
```
		errMsg := fmt.Sprintf("cannot find a network-attachment-definition (%s) in namespace (%s): %v", net.Name, net.Namespace, err)
		if try && apierrors.IsNotFound(err) {
			logging.Verbosef(errMsg)
			return nil, resourceMap, fmt.Errorf("getKubernetesDelegate: " + errMsg)
		}
```

Reply:
In such case, we expect that net-attach-def is created with empty. In your case, you need to create this net-attach-def.

```
---
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: cilium
spec:
```

Reply:
In addition, if you proceed to introduce this argument, I recommend to change the name `try` to something because `try` is not clear what it does. Anyway, I still prefer not to add new arguments for the purpose.

Reply:
Or you can just specify CNI path in `clusterNetwork`, your requirement might be fullfilled.

If you put cilium CNI config in '/etc/cni/net.d/10-cilium.conflist`, then 

```
{
  "name": "multus-cni-network",
  "type": "multus",
  "confDir": "/etc/cni/net.d",
...
  "clusterNetwork": "/etc/cni/net.d/10-cilium.conflist"
}
```

Reply:
I dont want to hard code the path `/etc/cni/net.d/10-cilium.conflist`. Any recommendation for the name?

Reply:
First, let's find out how to archieve that without change.

In your case, you can add net-attach-def with the name 'cilim' in namespace default, then it works, I suppose. Did you try that?

Reply:
I do not want to maintain the "cilium" NAD, which means I have to create it, handle updates/upgrades and ensure it is never deleted, so that option does not work for me.